[CP] Make context_parallel_shard more general

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #169393
* #170201
* __->__ #170200

**Summary:**

Fix buffer sharding to support batch dimensions created through expand() or view() operations.

The previous implementation used torch.index_select in a loop over the batch dimension, which fails when the batch dimension is created through expand() or view() (e.g., when a tensor with shape (seq_len,) is expanded to (batch_size, seq_len)). This is because each modification (assignment inside the loop) operates on the same storage.

**Changes:**
- Replace the per-batch-item loop using torch.index_select with a single torch.gather operation that handles the entire batch at once
- Add explicit validation to ensure buffers are 2D or less (shape (seq_len,) or (batch_size, seq_len))

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update on "[CP] Make context_parallel_shard more general"


**Summary:**

Fix buffer sharding to support batch dimensions created through expand() or view() operations.

The previous implementation used torch.index_select in a loop over the batch dimension, which fails when the batch dimension is created through expand() or view() (e.g., when a tensor with shape (seq_len,) is expanded to (batch_size, seq_len)). This is because each modification (assignment inside the loop) operates on the same storage.

**Changes:**
- Replace the per-batch-item loop using torch.index_select with a single torch.gather operation that handles the entire batch at once
- Add explicit validation to ensure buffers are 2D or less (shape (seq_len,) or (batch_size, seq_len))

[ghstack-poisoned]

tianyu-l commented on torch/distributed/tensor/experimental/_context_parallel/_attention.py:
```diff
@@ -1089,10 +1089,14 @@ def _context_parallel_buffers(
-
             # NOTE: assuming batch dim is 0
 
+            if len(buffer.shape) > 2:
```
why even have this restriction? My worry is that in multimodal there might be [bs, seqlen, other dims] inputs, but I don't have concrete use case.

fegin replied:
nvm, I made it work with multiple dimensions.

fegin committed:
Update on "[CP] Make context_parallel_shard more general"


**Summary:**

Fix buffer sharding to support batch dimensions created through expand() or view() operations.

The previous implementation used torch.index_select in a loop over the batch dimension, which fails when the batch dimension is created through expand() or view() (e.g., when a tensor with shape (seq_len,) is expanded to (batch_size, seq_len)). This is because each modification (assignment inside the loop) operates on the same storage.

**Changes:**
- Replace the per-batch-item loop using torch.index_select with a single torch.gather operation that handles the entire batch at once
- Add explicit validation to ensure buffers are 2D or less (shape (seq_len,) or (batch_size, seq_len))

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update

[ghstack-poisoned]
